### DISCUSSION TITLE
which type of comparer does mobx use ?

### DISCUSSION THREAD

**Discussion by prashantkn94** [mobxjs/mobx#3751] [Q&A]
I see there are different type of comparers in packages/mobx/src/utils/comparer.ts . But which type of comparison is actually done on the observables value before deciding whether to re-trigger render for React component.  
Also, what comparer is used for reactivity in general ?

**Comment by mweststrate** [comment]
reference equality, which is in 98% of the cases the right choice. Not 100%
sure whether that is `===` or `Object.is` from top of my head

**Reply by prashantkn94** [reply]
reference equality would fail in case the observable is an object if any value of the key of the object is changed

**Reply by mweststrate** [reply]
no it normally won't, make sure to read up on the docs first to see how reactivity works in MobX.

**Reply by prashantkn94** [reply]
I have read the docs on reactivity. Let's say we have an observable object like below

`@observable obj = { name: 'first', lastName": 'last' };`

and we update it like below:

`obj.name = "myName"`

Now the reference of the obj is still the same but mobx will re-trigger the render. Please explain

**Comment by mweststrate** [comment]
MobX traps read and writes to all properties of observable objects, so if
you read obj.name in observer, it will be detected and matched against
future mutations of that property. See
https://mobx.js.org/understanding-reactivity.html#understanding-reactivity
or the original blog post:
https://www.mendix.com/blog/making-react-reactive-pursuit-high-performing-easily-maintainable-react-apps/
